drv/sprot-api/src/lib.rs

```diff
@@ -3,64 +3,266 @@
+    pub blob: Option<BlobInfo>,
+}
+
+impl<T> Msg<T>
+where
+    T: Serialize + for<'a> Deserialize<'a> + SerializedSize,
+{
+    /// Serialize a `Header` followed by a `ReqBody` or `RspBody`, compute a CRC, serialize
+    /// the CRC, and return the total size of the serialized request.
+    pub fn pack(body: &T, buf: &mut [u8]) -> Result<usize, SprotProtocolError> {
+        // Serialize `body`
+        let mut size = hubpack::serialize(&mut buf[Header::MAX_SIZE..], body)?;
+
+        // Create a header, now that we know the size of the body
+        let header = Header::new(size.try_into().unwrap_lite());
+
+        // Serialize the header
+        size += hubpack::serialize(buf, &header)?;
+
+        // Compute and serialize the CRC
+        let crc = CRC16.checksum(&buf[..size]);
+        size += hubpack::serialize(&mut buf[size..], &crc)?;
+
+        Ok(size)
+    }
+
+    /// Serialize a `Header` followed by a `ReqBody` or `RspBody, copy a blob
+    /// into `buf` after the serialized body,  compute a CRC, serialize the
+    /// CRC, and return the total size of the serialized request.
+    pub fn pack_with_blob(
+        body: &T,
+        buf: &mut [u8],
+        blob: LenLimit<Leased<R, [u8]>, MAX_BLOB_SIZE>,
+    ) -> Result<usize, SprotProtocolError> {
+        // Serialize `body`
+        let mut size = hubpack::serialize(&mut buf[Header::MAX_SIZE..], body)?;
+
+        // Copy the blob into the buffer after the serialized body
+        blob.read_range(0..blob.len(), &mut buf[Header::MAX_SIZE + size..])
+            .map_err(|_| SprotProtocolError::ServerRestarted)?;
+
+        size += blob.len();
+
+        // Create a header, now that we know the size of the body
+        let header = Header::new(size.try_into().unwrap_lite());
+
+        // Serialize the header
+        size += hubpack::serialize(buf, &header)?;
+
+        // Compute and serialize the CRC
+        let crc = CRC16.checksum(&buf[..size]);
+        size += hubpack::serialize(&mut buf[size..], &crc)?;
+
+        Ok(size)
+    }
+
+    // Deserialize and return a `Msg`
+    pub fn unpack(buf: &[u8]) -> Result<Msg<T>, SprotProtocolError> {
+        let (header, body_buf) = hubpack::deserialize::<Header>(buf)?;
+        if header.protocol != Protocol::V2 {
+            return Err(SprotProtocolError::UnsupportedProtocol);
+        }
+        Self::unpack_body(header, buf, body_buf)
+    }
+
+    /// Deserialize just the body, given a header that was already deserialized.
+    pub fn unpack_body(
+        header: Header,
+        // The buffer containing the entire serialized `Msg` including the `Header`
+        buf: &[u8],
+        // The body part of the buffer including the CRC at the end
+        body_buf: &[u8],
+    ) -> Result<Msg<T>, SprotProtocolError> {
+        let (body, blob_buf) = hubpack::deserialize::<T>(body_buf)?;
+        let end = Header::MAX_SIZE + header.body_size as usize;
+        let blob_len =
```

Comment:
(note that if you make blob into a `&[u8]` as I suggested above, the rest of this code...goes away, because the correct slice is already available as `blob_buf`)